client/session.go

```diff
@@ -758,30 +763,41 @@ func (s *session) streamBlocksMetadataFromPeer(
 		}
 
 		for _, elem := range result.Elements {
-			blocks := make([]blockMetadata, len(elem.Blocks))
-			for i, b := range elem.Blocks {
-				blocks[i].start = time.Unix(0, b.Start)
+			blockMetas := make([]blockMetadata, 0, len(elem.Blocks))
+			for _, b := range elem.Blocks {
+				var (
+					blockStart = time.Unix(0, b.Start)
+					blockEnd   = start.Add(blockSize)
+				)
+				if start.After(blockEnd) || !blockStart.Before(end) {
```

Comment:
nit: I think `blockEnd` is exclusive, so could do `!start.Before(blockEnd) || !blockStart.Before(end)`